### PR TITLE
Implement configurable caching

### PR DESCRIPTION
**Pull request by Phil1602**
Hi,

thanks for this exporter.

We have many Promethei, which are scraping in parallel with a small scrape interval. Unfortunately we discovered load issues because of the DellHW Exporter respectively the expensive underlying omreport calls (all collectors were enabled) on our machines.

We decided to implement a simple in memory cache, which can be enabled using the configuration flags, to face this problem. The caching is optional and can be enabled using the configuration parameters.

**Implementation details:**
I decided to implement an additional adapter channel, which will be used to retrieve the collected metrics and put them into an array if caching is enabled. A mutex is used to prevent concurrent collections and concurrent write operations to the cache.

Since the metrics are written into a "local" channel instead of the channel passed by the Prometheus library directly, we need a second waitgroup. The first waitgroup ensures that all collectors have finished. The second waitgroup ensures that all metrics are written to the outgoing channel before the method returns. This is needed because the Prometheus library will close the channel once the method returns.

There was already a field for the last collection timestamp, which was not actually used before. The initial timestamp will now be Unix 0 instead of the current time (start time of the exporter), because otherwise the actual collection would not be executed within the first cache period if caching is enabled. Also the receiver had to be changed  to Pointer instead of Value, because the updated timestamp was only updated for the copy in case of the value receiver.

The new configuration parameters have been added to the configuration documentation. Shall I add some more documentation to further describe the caching behavior for the end user? What about the debug logs, are they ok or shall I remove them?

Thanks!
Philipp

### CODE REVIEW

File: cmd/dellhw_exporter/dellhw_exporter.go
Comment (galexrt) [inline cmd/dellhw_exporter/dellhw_exporter.go:85]:
```suggestion
}

func init() {
```

Please add an empty line between the function.

**Review by galexrt — CHANGES_REQUESTED** [review]
Thanks for the PR!

Adding the cache implementation details as a new page to the docs sounds good.
I might "move" it around in the doc structure later on, but don't worry about that.

I only have one small comment regarding the code.

File: cmd/dellhw_exporter/dellhw_exporter.go
Comment (Phil1602) [reply cmd/dellhw_exporter/dellhw_exporter.go]: Good catch, thanks!
Unfortunately `go fmt` did not caught it.

**Comment by Phil1602** [comment]
I will add the documentation page tomorrow.

File: README.md
Comment (galexrt) [inline README.md:40]:
```suggestion
## Caching

Optional caching can be enabled to prevent performance issues caused by this exporter, see [Caching doc page](docs/caching.md).
```

File: docs/caching.md
Comment (galexrt) [inline docs/caching.md:4]:
```suggestion
## Caching

Why do we even need caching in an exporter?
```

File: docs/caching.md
Comment (galexrt) [inline docs/caching.md:25]:
```suggestion
### Implementation details

An additional adapter channel is used to retrieve the collected metrics and put them into an array if caching is enabled. A mutex is used to prevent concurrent collections and concurrent write operations to the cache.
```

**Review by galexrt — CHANGES_REQUESTED** [review]
LGTM, just some nit picks regarding the docs.
I'll merge after those have been fixed.

**Comment by Phil1602** [comment]
Sorry I didn't had time yesterday.
The documention has been added.

**Comment by Phil1602** [comment]
Applied your suggestions and squashed the documentation related commits together.

**Review by galexrt — APPROVED** [review]

**Comment by galexrt** [comment]
@Phil1602 Thanks for the PR and quick updates!

I'll push a new release in a few hours after the merge.

**Comment by galexrt** [comment]
@Phil1602 [v1.7.0 has been released.](https://github.com/galexrt/dellhw_exporter/releases/tag/v1.7.0)

**Pull request merged**
